dff/context_storages/context_schema.py

```diff
@@ -0,0 +1,270 @@
+        Calculate what fields to write, split large data into chunks if needed and call writer function.
+        Also update `updated_at` attribute of the given context with current time, set `primary_id` and `storage_key`.
+
+        :param ctx: the context to store.
+        :param pac_writer: the function used for writing context to `CONTEXT` table (see :py:const:`~._WritePackedContextFunction`).
+        :param log_writer: the function used for writing context to `LOGS` table (see :py:const:`~._WriteLogContextFunction`).
+        :param storage_key: the key to store the context with.
+        :param chunk_size: maximum number of items that can be inserted simultaneously, False if no such limit exists.
+
+        :return: the read :py:class:`~.Context` object.
+        """
+        updated_at = datetime.now()
+        setattr(ctx, ExtraFields.updated_at.value, updated_at)
+        created_at = getattr(ctx, ExtraFields.created_at.value, updated_at)
+
+        ctx_dict = ctx.dict()
+        logs_dict = dict()
+        primary_id = getattr(ctx, ExtraFields.primary_id.value, str(uuid4()))
+
+        for field_props in [value for value in dict(self).values() if isinstance(value, SchemaField)]:
+            nest_dict = ctx_dict[field_props.name]
+            last_keys = sorted(nest_dict.keys())
+
+            if self.append_single_log and isinstance(field_props.subscript, int) and len(nest_dict) > field_props.subscript:
+                unfit = -field_props.subscript - 1
+                logs_dict[field_props.name] = {last_keys[unfit]: nest_dict[last_keys[unfit]]}
+            else:
+                if self.duplicate_context_in_logs or not isinstance(field_props.subscript, int):
+                    logs_dict[field_props.name] = nest_dict
+                else:
+                    logs_dict[field_props.name] = {key: nest_dict[key] for key in last_keys[:-field_props.subscript]}
+
+            if isinstance(field_props.subscript, int):
+                last_keys = last_keys[-field_props.subscript:]
+
+            ctx_dict[field_props.name] = {k: v for k, v in nest_dict.items() if k in last_keys}
+
+        await pac_writer(ctx_dict, created_at, updated_at, storage_key, primary_id)
+
+        flattened_dict: List[Tuple[str, int, Dict]] = list()
+        for field, payload in logs_dict.items():
+            for key, value in payload.items():
+                flattened_dict += [(field, key, value)]
+        if len(flattened_dict) > 0:
+            if not bool(chunk_size):
+                await log_writer(flattened_dict, updated_at, primary_id)
+            else:
+                tasks = list()
+                for ch in range(0, len(flattened_dict), chunk_size):
+                    next_ch = ch + chunk_size
+                    chunk = flattened_dict[ch:next_ch]
+                    tasks += [log_writer(chunk, updated_at, primary_id)]
+                if self.supports_async:
+                    await gather(*tasks)
```

Comment:
I have no idea myself, but is it safe to write chunks of the same data in parallel without locks/queues?

Reply:
The general idea is that `supports_async` flag, if set, means "we shouldn't bother about asynchronous requests to database, it's capable of managing them itself".
Some databases (e.g. PostgreSQL, MongoDB, etc.) can manage parallel reads and writes themselves, for them we can send all the data simultaneously. In cases we're not sure that database can handle async requests, we set `supports_async` to False and in that case all requests will be synchronous.
P.S. The data is not the same, all chunks contain different data pieces.